src/components/implementation/no_interface/cFE_booter/osfilesys.h

```diff
@@ -1,113 +1,258 @@
+#define F_PART_DATA_SIZE (4096-sizeof(struct f_part))
 
-//should be overwritten by linking step in build process
-__attribute__((weak)) char _binary_cFEfs_tar_size=0;
-__attribute__((weak)) char _binary_cFEfs_tar_start=0;
-__attribute__((weak)) char _binary_cFEfs_tar_end=0;
+// currently we compile composite and cFE with different libc
+// muslc for cos and whatever ubuntu32 14 use for cFE
+// these have different structs so I need to implement a dirent struct in cos
+// with the same layout as cFE's dirent
+struct hack_dirent
+{
+    ino_t d_ino;
+    // off_t d_off; these are the 8 bytes that I am removing
+    unsigned short d_reclen;
+    unsigned char d_type;
+    char d_name[256];
+};
 
-//locations and size of tar
-char      *start;
-char      *end;
-size_t    size;
+// for exactly the same reason as above, we need to declare a stat structure to pass
+// back to the cFE that it will recoginize
+struct hack_stat
+{
+    int64 st_dev;
+    uint32 padding1; // here to keep the offset consistant
+    int32 st_ino;
+    int32 st_mode;
+    int32 st_nlink;
+    int32 st_gid;
+    int32 st_uid;
+    int64 st_rdev;
+    uint32 padding2; // also here to keep offset consistant with cFE
+    int32 st_size;
+    int32 st_blksize;
+    int32 st_blocks;
+    int32 st_hack_atime; // st_time field are defined elsewhere
+    int32 st_hack_mtime;
+    int32 st_hack_ctime;
+};
 
 typedef enum {
     FSOBJ_FILE,
     FSOBJ_DIR,
-} fsobj_type_t;
+} fsobj_type; //TODO: what is the preferred naming convention for this
+
+typedef enum {
+    READ,
+    WRITE,
+    READ_WRITE,
+    NONE,
+} permission_t;
+
+struct file_position {
+    struct f_part *open_part;  //part of file being read/written to
+    uint32 part_offset;         //offset into part's data
+    uint32 file_offset;         //position within file as a whole
+};
+
+typedef enum {
+    STATIC,
+    DYNAMIC,
+} fpart_alloc_t;
+
+/*
+ * The structure of the filesystem is stored in the fsobj struct
+ * each non-leaf dir points to its first child, which then has a pointer
+ * to the next, etc. Each file has a linked list of f_parts to store blocks of data
+ */
 
 struct fsobj {
     char *name;
-    int32 filedes; // 0 for free file, posative for tracked file
-    fsobj_type_t type;
+    int32 ino; // 0 for free file
+    fsobj_type type; /* dir vs file, determines the type of FD position */
     size_t size;
-    uint32 refcnt; //number of programs which have opened it
-    char *data;
+    uint32 refcnt; //number of filedes which have it opened
+    permission_t permission; // most permissive possible status it may be opened with
+    fpart_alloc_t memtype;
+    struct f_part *file_part;
     struct fsobj *next, *prev;
-    struct fsobj *child, *parent;   /* child != NULL iff type = dir */
+    struct fsobj *child, *parent; // child != NULL iff type = dir
+};
+
+struct f_part {
+    struct fsobj *file;
+    struct f_part *next, *prev;
+    char *data;
 };
```

Comment:
true, but that is not something I want to change in this PR.  I think it is still worth putting into the repo before we switch to PS list